Add support for custom AWS profiles.

niall-mediasuite opened the pull request:
This change allows specification of the AWS profile to use for code artefact logins. At the moment this only supports a single value but in future could be expanded to make the selection based on the input URIs domain / account / region.

jmkeyes commented on keyrings/codeartifact.py:
```diff
@@ -1,18 +1,38 @@
 #!/usr/bin/env python
 
+import profile
```
I'm fairly certain this isn't used anywhere. Is this import necessary?

niall-mediasuite replied:
Woops, sorry. I should have paid closer attention. I think VSCode auto imported must have done this at some point and I didn't pick up on it. I'll strip it out.

jmkeyes commented on keyrings/codeartifact.py:
```diff
@@ -1,18 +1,38 @@
 
 from keyring import backend
 from keyring import credentials
+from keyring.util import platform_ as platform
```
The `keyring.util.platform_` package has been suffixed with an underscore to disambiguate it from the Python 3 built-in `platform` package. Instead of shadowing that built-in package, we should import only the function we need:

```python
from keyring.util.platform_ import config_root
```

niall-mediasuite replied:
Fair enough. I typically like to import modules rather than putting functions into the global namespace but its totally a preference. I'll update this to be consistent with the other imports.

jmkeyes commented on keyrings/codeartifact.py:
```diff
@@ -53,8 +73,12 @@ def get_password(self, service, username):
+
         # Create a CodeArtifact client for this repository's region.
-        client = boto3.client('codeartifact', region_name=region)
+        session = boto3.Session(profile_name=profile_name)
```
The use of a [`boto3.Session`](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/session.html#session-configurations) here is a good idea, but this implementation only allows for setting the profile name.

As an alternative (though I haven't tested this yet):

```python
# ...
import configparser

# ...
from functools import cache


# Allowed options for this backend.
_INI_OPTIONS = {
  "aws_access_key_id",
  "aws_secret_access_key",
  # "region_name",
  "profile_name",
}


@cache
def _load_config():
  keyring_config_file = config_root() / 'keyringrc.cfg'

  if not os.path.exists(keyring_config_file):
    return {}

  config = configparser.RawConfigParser()
  config.read(keyring_config_file)

  cfg = {}
  for option in _INI_OPTIONS:
    try:
      cfg[option] = config.get("codeartifact", option)
    except (configparser.NoSectionError, configparser.NoOptionError):
      pass

  return cfg

```

And within the `CodeArtifactBackend`'s `get_password()` method:
```python
  # Load our configuration file for the session.
  session_config = _load_config()
  session = boto3.Session(**session_config)

  # Create a CodeArtifact client for this repository's region.
  client = session.client('codeartifact', region_name=region)
```

Now the backend will react to any allowed options within the `[codeartifact]` section (if it exists) within the `keyringrc.cfg` configuration file.

niall-mediasuite replied:
Yeah nice idea. I just added the option I needed but I'm happy to give this a quick test and make the adjustment.

jmkeyes reviewed (changes requested):
Thank you for this contribution! I think this is a great idea. 

I've left some comments for your review; let me know what you think!

niall-mediasuite commented:
Thanks, I've updated based on your comments and updated the documentation to match. Please let me know if there are any other changes you want me to make.

niall-mediasuite commented on keyrings/codeartifact.py:
```diff
@@ -1,18 +1,49 @@
+import configparser
 
 import boto3
-import botocore
```
Also discovered another import that doesn't appear to be used.

jmkeyes replied:
Good catch!

jmkeyes commented:
Made some minor cosmetic changes. I'll squash and merge.

jmkeyes commented:
Version 1.1.0 has been built and pushed to PyPi.

niall-mediasuite commented:
Thank you.
